feat: introduce strong types for cloud types

tlm opened the pull request:
This commit introduces strong types for representing the type of a cloud. Currently when talking about a cloud's type inside that of a domain we use the string values and have no way to refer to a cloud type by a stronger primitive.

In this commit the cloud domain now has a CloudType value that is based off of the integer index in the DDL. The core/cloud package now also includes a set of CloudType enums that are based off of the string representation of the cloud type.

Tests have been added to the cloud domain to check first that the integer enums in the domain are aligned with the values that exist within the controller DDL. Another test in the cloud domain also checks that the core cloud type string values align with the type string representation of a cloud type in the controller DDL.

On top of this tests have been added to each provider to make sure that the provider type value each provider uses to register itself in the environ registry is the same as the value defined in core/cloud. This now means all the way from the the DDL to the provider we have unit tests to detect skew and drift of values.

The tests for the provider were done from the providers aspect of the problem. If the test was to be conducted from any other part of the code base it would bring the provider package in as a dependency. This would introduce the problem where we are unable to compile out certain providers and their dependencies from the controller.

The work here is done in advance to support a new ModelManager domain that is being developed. The ModelManager domain needs to be able to look at the cloud type of a model and calculate a corresponding type for the model. By now having this logic rely on the integer cloud type enums it means less lookup and matching needs to be performed within the database.

This pattern also makes the cloud domain and cloud type more inline with the idiomatic approach Juju is going in with DQlite where we prefer unique identifiers over that of strings.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
-  [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Unit tests validate this unused feature for the moment. Future model manager PR's will start making use of this.

## Documentation changes

N/A

## Links

hmlanigan commented on domain/cloud/cloudtype_test.go:
```diff
@@ -0,0 +1,92 @@
+// [corecloud.CloudType]. If this tests fails it means the database has a
+// disconnect with the core cloud types and this needs to be corrected to be
+// able to establish providers correctly in the controller.
+func (s *cloudTypeSuite) TestCloudTypeDBValuesAgainstCoreCloudTypes(c *tc.C) {
```
💯

hmlanigan commented on core/cloud/cloudtype.go:
```diff
@@ -0,0 +1,46 @@
+	// CloudTypeOpenStack represents the OpenStack cloud provider.
+	CloudTypeOpenStack CloudType = "openstack"
+
+	// CloudTypeVSphere represents the vSphere cloud provider.
+	CloudTypevSphere CloudType = "vsphere"
```
nit: the comment and the type have slightly different spelling. Capital vs lowercase `v`.

`CloudTypeVSphere` is easier to read for me.

tlm replied:
yeah I was  trying to keep things inline with the offical product name where I could. In hind sight probably just makes the problem that much worse. Will get this changed.

tlm commented:
@hmlanigan and @wallyworld,

I have made some changes to this PR to address some recent design discussions that have happened. The real change is removing the core cloud types that I had previously added. Because we want to move out of core I don't want to be dragging more new work into the package.

Instead I have added a `String()` method to the domain type to get the string representation. These values are tested against that of the providers to check alignment. Have also introduced a means by upper and lower bounds on the int enum type to check if a value is valid and also detect more skew cases in the database.

The change is still in line with the original PR I have just removed the additions to core.

tlm commented:
/merge
